### ISSUE TITLE
SAVEPOINT not released on ROLLBACK in PostgreSQL

### ISSUE THREAD

**Issue by dgobaud** [rails/rails#22404]
In PostgreSQL when ROLLBACK is done on a SAVEPOINT http://www.postgresql.org/docs/9.1/static/sql-rollback-to.html it does not RELEASE it http://www.postgresql.org/docs/9.1/static/sql-release-savepoint.html so it continues to consume resources. There are cases where you want this behavior but I don't think the behavior can be obtained using ActiveRecord so I think it should also do a RELEASE on ROLLBACK. Code to see this is below along with SQL log output. The example where you would not want a SAVEPOINT release is for example

```
savepoint foo;
A;
if A failed then
  rollback to foo;
  B;
  if B failed then
    rollback to foo;
    C;
    .....
release foo;
```

I don't think this can be done with ActiveRecord.

```
ActiveRecord::Base.transaction do
  begin
    ActiveRecord::Base.transaction(:requires_new => true) do
      raise 'ahhh'
    end
  rescue
  end
end
```

   (0.3ms)  BEGIN
   (0.2ms)  SAVEPOINT active_record_1
   (0.2ms)  ROLLBACK TO SAVEPOINT active_record_1
   (0.2ms)  COMMIT

**Comment by RochesterinNYC** [comment]
I've been working on a PR to provide configuration opt-out of this behavior. Will reference this issue when I submit the PR.

**Comment by arthurnn** [comment]
@RochesterinNYC cc me on the PR, thanks

**Comment by dgobaud** [comment]
@RochesterinNYC any idea if this got done or merged into 5.0? thanks!
